### PR TITLE
Enhance for with additional delayed predicate function

### LINKED ISSUE #5800: make verify fails on Macbook with ARM64 architecture family
**How to categorize this issue?**

/area dev-productivity
/area testing
/kind bug

**What happened**:
On Macbook with:
`$(go env GOOS) = darwin`  && `$(go env GOARCH) = arm64`
when you run `make verify` then it fails with the following error log:
```
> Installing promtool
Unknown architecture
make: *** [hack/tools/bin/promtool] Error 255
```
**What you expected to happen**:
`make verify` should succeed when run on Macbook M1 chipset which belongs to `ARM64` architecture family.

**How to reproduce it (as minimally and precisely as possible)**:
Just run `make verify` on Macbook M1.

**Anything else we need to know?**:

The version of promtool used currently is `2.24.1` which does not have an archive for `darwin` && `arm64`. Version `2.26.0` onwards an archive for the same has been added. The latest version of the `promtool` is `2.34.0`

In the script there is a snippet:
```bash
case $(uname -m) in
  aarch64)
    arch="arm64"
    ;;
  x86_64)
    arch="amd64"
    ;;
  *)
    echo "Unknown architecture"
    exit -1
    ;;
esac
```
When you execute `uname -m` on macbook M1 terminal then you directly get `arm64` so it does not match any case and gives `Unknown architecture`. This also needs to be corrected.

**Environment**:
`uname -a` >> `Darwin CP2H3HJQML 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:46:32 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T6000 arm64`

- Gardener version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### PR DESCRIPTION
**Pull request by unmarshall**
**How to categorize this PR?**
/area control-plane
/kind enhancement

**What this PR does / why we need it**:
Introduces a new function `DoIfFn` in `TaskFn` allowing delayed evaluation of a predicate.

**Which issue(s) this PR fixes**:
Fixes Part of #5710
Fixes #5800

**Special notes for your reviewer**:
As discussed in the issue #5710 the remainder of the proposals can be dropped unless there is a consensus to put them in as convenience functions.

**Release note**:

### CODE REVIEW

File: pkg/utils/flow/taskfn.go [hunk 1]
```diff
@@ -53,6 +53,12 @@ func (t TaskFn) DoIf(condition bool) TaskFn {
 	return t.SkipIf(!condition)
 }
 
+// DoIfFn returns a TaskFn that will be executed if the invocation of conditionalFn is true when it is called.
+// Otherwise, it will do nothing when called.
+func (t TaskFn) DoIfFn(conditionalFn func() bool) TaskFn {
+	return t.SkipIf(!conditionalFn())
```
Comment (rfranzke) [inline pkg/utils/flow/taskfn.go:59]: Why not `t.DoIf(conditionalFn())` to make it easier readable/straight-forward? 😕

Comment (unmarshall) [reply pkg/utils/flow/taskfn.go:59]: I wanted to construct the graph and compile it to a Flow without evaluating predicates during construction phase. And re-use the compiled flow repeatedly using different closure states when its run. Essentially its a lazy evaluation of a predicate. Let me test this again just to confirm that this is then possible using existing `DoIf`

Comment (unmarshall) [reply pkg/utils/flow/taskfn.go:59]:
```go
		It("should process the function only when the conditional function returns true", func() {
			results := make([]string, 0, 2)
			var trueFnEvaluated bool
			var falseFnEvaluated bool
			trueFn := func() bool {
				trueFnEvaluated = true
				return true
			}
			falseFn := func() bool {
				falseFnEvaluated = true
				return false
			}
			var (
				g = flow.NewGraph("foo")
				_ = g.Add(flow.Task{Name: "a-task", Fn: flow.TaskFn(func(ctx context.Context) error {
					results = append(results, "a-task-result")
					return nil
				}).DoIf(trueFn())})
				_ = g.Add(flow.Task{Name: "b-task", Fn: flow.TaskFn(func(ctx context.Context) error {
					results = append(results, "b-task-result")
					return nil
				}).DoIf(falseFn())})
				f = g.Compile()
			)
			Expect(trueFnEvaluated).To(BeFalse())
			Expect(falseFnEvaluated).To(BeFalse())
			Expect(f.Run(ctx, flow.Opts{})).ToNot(HaveOccurred())
			Expect(results).To(HaveLen(1))
			Expect(results[0]).To(Equal("a-task-result"))
		})

```
In the test above you can see that the predicate functions are actually evaluated at construction time and before flow.Run is called. This is not what was desired. After a bit more testing i realize now that introducing DoIfFn also does not cause lazy evaluation of predicate in its current form.

Comment (rfranzke) [reply pkg/utils/flow/taskfn.go:59]: I cannot follow. My suggestions was only to change 

```suggestion
	return t.DoIf(conditionalFn())
```

to make it easier readable. 

Also, can we introduce a similar `SkipIfFn` for symmetry reasons?

Also, please check https://github.com/gardener/gardener/pull/5801#issuecomment-1097725099

Comment (unmarshall) [reply pkg/utils/flow/taskfn.go:59]: Ok the only way to have a lazy evaluation of a predicate is to have the predicate as part of the TaskFn itself. I will abandon this PR as the proposed change anyways does not solve what i originally intended. Should have done better testing/validation of the intent. My bad.

**Comment by unmarshall** [comment]
The change does not resolve the issue. Abandoning it.

**Pull request closed without merging**